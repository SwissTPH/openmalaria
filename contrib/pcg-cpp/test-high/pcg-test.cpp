/*
 * PCG Random Number Generation for C++
 *
 * Copyright 2014-2017 Melissa O'Neill <oneill@pcg-random.org>,
 *                     and the PCG Project contributors.
 *
 * SPDX-License-Identifier: (Apache-2.0 OR MIT)
 *
 * Licensed under the Apache License, Version 2.0 (provided in
 * LICENSE-APACHE.txt and at http://www.apache.org/licenses/LICENSE-2.0)
 * or under the MIT license (provided in LICENSE-MIT.txt and at
 * http://opensource.org/licenses/MIT), at your option. This file may not
 * be copied, modified, or distributed except according to those terms.
 *
 * Distributed on an "AS IS" BASIS, WITHOUT WARRANTY OF ANY KIND, either
 * express or implied.  See your chosen license for details.
 *
 * For additional information about the PCG random number generation scheme,
 * visit http://www.pcg-random.org/.
 */

/*
 * This file is based on the demo program for the C generation schemes.
 * It shows some basic generation tasks.
 */

#include <cstddef>
#include <cstdint>
#include <cstdlib>
#include <cassert>
#include <climits>
#include <iostream>
#include <sstream>
#include <iomanip>
#include <algorithm>
#include <numeric>
#include <random>       // for random_device

#include "pcg_random.hpp"

// This code can be compiled with the preprocessor symbol RNG set to the
// PCG generator you'd like to test.

#ifndef RNG
    #define RNG pcg32
    #define TWO_ARG_INIT 1
#endif

#define STRINGIFY_IMPL(x) #x
#define STRINGIFY(x)      STRINGIFY_IMPL(x)

using namespace std;
using pcg_extras::operator<<;

#if !PCG_EMULATED_128BIT_MATH || !AWKWARD_128BIT_CODE

int main(int argc, char** argv)
{
    // Read command-line options

    int rounds = 5;
    bool nondeterministic_seed = false;

    ++argv;
    --argc;
    if (argc > 0 && strcmp(argv[0], "-r") == 0) {
        nondeterministic_seed = true;
        ++argv;
        --argc;
    }
    if (argc > 0) {
        rounds = atoi(argv[0]);
    }

    /* Many of the generators can be initialized with two arguments; the second
     * one specifies the stream.
     */

#if TWO_ARG_INIT
    RNG rng(42u, 54u);
#else
    RNG rng(42u);
#endif

    if (nondeterministic_seed) {
        // Seed with external entropy from std::random_device (a handy
        // utility provided by pcg_extras).
        rng.seed(pcg_extras::seed_seq_from<std::random_device>());
    }

    constexpr auto bits = sizeof(RNG::result_type) * CHAR_BIT;
    constexpr int how_many_nums = bits <= 8  ? 14
                                : bits <= 16 ? 10
                                :              6;
    constexpr int wrap_nums_at  = bits >  64 ? 2
                                : bits >  32 ? 3
                                :              how_many_nums;

    cout << STRINGIFY(RNG) << ":\n"
    //   << "      -  aka:         " << pcg_extras::printable_typename<RNG>()
    // ^-- we skip this line because the output is long, scary, ugly, and
    //     and varies depending on the platform
         << "      -  result:      " <<  bits << "-bit unsigned int\n"
         << "      -  period:      2^" << RNG::period_pow2();
    if (RNG::streams_pow2() > 0)
         cout << "   (* 2^" << RNG::streams_pow2() << " streams)";
    cout << "\n      -  size:        " << sizeof(RNG) << " bytes\n\n";

    for (int round = 1; round <= rounds; ++round) {
        printf("Round %d:\n", round);

        /* Make some N-bit numbers */
        cout << setw(4) << setfill(' ') << bits << "bit:";
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << endl;

        cout << "  Again:";
        rng.backstep(6);
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << dec << endl;

        /* Toss some coins */
        cout << "  Coins: ";
        for (int i = 0; i < 65; ++i)
            cout << (rng(2) ? "H" : "T");
        cout << endl;

        RNG rng_copy{rng};
        /* Roll some dice */
        printf("  Rolls:");
        for (int i = 0; i < 33; ++i)
            cout << " " << (uint32_t(rng(6)) + 1);
        cout << "\n   -->   rolling dice used "
             << static_cast<size_t>(rng - rng_copy)
             << " random numbers" << endl;

        /* Deal some cards using pcg_extras::shuffle, which follows
         * the algorithm for shuffling that most programmers would expect.
         * (It's unspecified how std::shuffle works.)
         */      
        enum { SUITS = 4, NUMBERS = 13, CARDS = 52 };
        char cards[CARDS];
        iota(begin(cards), end(cards), 0);
        pcg_extras::shuffle(begin(cards), end(cards), rng);

        /* Output the shuffled deck */
        printf("  Cards:");
        static const signed char number[] = {'A', '2', '3', '4', '5', '6', '7',
                                             '8', '9', 'T', 'J', 'Q', 'K'};
        static const signed char suit[] =   {'h', 'c', 'd', 's'};
        int i = 0;
        for (auto card : cards) {
            ++i;
            cout << " " << number[card / SUITS] << suit[card % SUITS];
            if (i % 22 == 0)
                cout << "\n\t";
        }
        
        cout << "\n" << endl;
    }

    // Ensure that the input/output functions work.
    std::stringstream ss;
    ss << rng;
    RNG from_string;
    ss >> from_string;
    assert(from_string == rng);

    return 0;
}

#else //  i.e. PCG_EMULATED_128BIT_MATH && AWKWARD_128BIT_CODE

int main()
{
    // It's not that it *can't* be done, it just requires either C++14-style
    // constexpr or some things not to be declared const.
    cout << "Sorry, " STRINGIFY(RNG) " not supported with emulated 128-bit math"
         << endl;
}

#endif
