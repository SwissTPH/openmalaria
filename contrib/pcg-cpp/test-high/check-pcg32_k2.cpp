#define RNG pcg32_k2
#define TWO_ARG_INIT 1

#include "pcg-test.cpp"

