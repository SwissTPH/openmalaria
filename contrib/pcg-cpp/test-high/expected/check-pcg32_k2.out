pcg32_k2:
      -  result:      32-bit unsigned int
      -  period:      2^128   (* 2^63 streams)
      -  size:        24 bytes

Round 1:
  32bit: 0x23ad45d5 0x6e2b9c53 0x23cf9e33 0x24e90350 0x7a160dc4 0x5cfeb7ad
  Again: 0x23ad45d5 0x6e2b9c53 0x23cf9e33 0x24e90350 0x7a160dc4 0x5cfeb7ad
  Coins: TTTTHTTTTHTTHTHTTHTTHTTTTTTTTHHHTHHTHTHTTHTHHTTHTHTTTHTHTHTTTHHHT
  Rolls: 3 4 4 6 5 5 5 6 6 1 6 6 6 3 4 5 6 5 3 3 6 4 5 2 1 2 1 1 4 4 4 2 5
   -->   rolling dice used 33 random numbers
  Cards: 8c 9s 4d 5s As 9c 2s Tc 7c Ks 5d Td 2d 7h 8h 2h Ah 6s 2c Qs Jd Kc
	 4c 8s Jc 5h 6d 4h 6c Kh Th Kd Js 9d 7d Qd 6h Qc Jh 3s 9h Ac 3c Ts
	 7s Qh 5c Ad 3h 4s 8d 3d

Round 2:
  32bit: 0x76ce5fd8 0x93a10635 0x3dcb594c 0x927f49c7 0x4ab6117c 0x05f1e67d
  Again: 0x76ce5fd8 0x93a10635 0x3dcb594c 0x927f49c7 0x4ab6117c 0x05f1e67d
  Coins: THTHTHHHTHHHHHHHHHHTHTTHHTHHTTTHHHHTTHHTHHTHTHTTTTHHHHHTHTHTHTHTH
  Rolls: 3 1 2 2 1 1 6 1 2 2 4 3 2 5 5 5 1 1 3 5 3 1 2 1 5 6 3 1 4 5 2 5 3
   -->   rolling dice used 33 random numbers
  Cards: 9h As 3d Tc 8h 5c 5d Jh 2s 6d 8s 4h 6c 9c 4s 4d Ts 2c 5h Ah Qh Js
	 Td Jc 3h 9s 3s Th Kd 8c Qs 2h Kh Qc 7h 6h 4c Ac 7s Qd 8d Ad 9d 7c
	 Kc 6s Ks 7d 2d 3c 5s Jd

Round 3:
  32bit: 0x5a9adcab 0x5d30c50f 0xfc94015f 0x3459e46f 0xc133138c 0x419a4672
  Again: 0x5a9adcab 0x5d30c50f 0xfc94015f 0x3459e46f 0xc133138c 0x419a4672
  Coins: HHTHTHHHTTHTTTTHHTTHHHHTTTTTTHTHHTTHTTHHHTTHHHTTHTTTHTTTTHTTTHTHT
  Rolls: 3 6 3 1 4 3 6 3 1 6 6 2 4 4 5 2 4 3 2 6 1 2 5 4 5 2 6 5 3 2 2 6 4
   -->   rolling dice used 33 random numbers
  Cards: 7d Ac 3c 7c Tc 4h Ad 9d 3s 6d 2h 7s 3d 5d Kh Th 8s 8d 4s 2d 4d Qc
	 3h Js 5s Jc 6c 2s Qs 5c Ks Jd Jh Ts Kd 8c 7h Qh 2c 4c As Qd 9s 8h
	 6h 6s 9h 9c Td Kc Ah 5h

Round 4:
  32bit: 0x33e94d75 0x25b26823 0x9210e5fa 0x5dd9e791 0xf2c25d15 0x3e8c34d3
  Again: 0x33e94d75 0x25b26823 0x9210e5fa 0x5dd9e791 0xf2c25d15 0x3e8c34d3
  Coins: TTHTHHHHTHTHHTTHTTHTTHHHHHHHHHTTHHHHTHTHTTTHHHTTTHHTHTTHHHTTHHTTT
  Rolls: 5 1 4 3 2 5 6 6 2 6 5 6 5 5 6 3 4 5 4 2 4 1 6 2 6 6 1 1 5 5 5 5 1
   -->   rolling dice used 33 random numbers
  Cards: Kh 6h 6d Ks 7c 7s 2c Qd Qc 5c 8s Th 8h 2s Td Ts As Jd 9c 4h Tc Qh
	 8c Ad 5h Jh 6c 4s 7d 2d 2h Kd 5d Ah 8d Kc 4d 9s 9d Js 4c Ac 3d Qs
	 5s 3s 7h 6s Jc 9h 3h 3c

Round 5:
  32bit: 0xb8130a51 0x3ffeab8a 0x38821a59 0x5a2600f2 0x3b6f403f 0x8117a786
  Again: 0xb8130a51 0x3ffeab8a 0x38821a59 0x5a2600f2 0x3b6f403f 0x8117a786
  Coins: TTHHTTTHHTHHTHHTTTTTTHTHHHTTTTTHTTHTTTTTTHHHTTTTHHHHHTHHHHHHTHHHH
  Rolls: 3 3 6 1 6 6 2 6 5 3 5 2 6 2 4 3 5 3 5 4 3 4 6 4 1 6 4 1 5 2 6 1 6
   -->   rolling dice used 33 random numbers
  Cards: Kh 4d Qs 5s Ks As 7s 2d 4h Qd Kc 6s 7d 3h Ad Jc 5h Js Ah 8c 6h 6c
	 Jd 9h 2h 2s 5d Kd 9c 8d 4c Tc 8h Jh 3c 8s 3d 7h 9d 3s 6d 5c Ts 4s
	 Th Ac 2c Td Qc Qh 9s 7c

